cranelift/codegen/src/isa/aarch64/pcc.rs

```diff
@@ -113,37 +263,32 @@ fn check_addr(
             ty,
             extendop,
         } => {
-            let rn = vcode.vreg_fact(rn.into()).ok_or(PccError::MissingFact)?;
-            let rm = vcode.vreg_fact(rm.into()).ok_or(PccError::MissingFact)?;
-            let rm_extended = amode_extend(ctx, rm, extendop).ok_or(PccError::MissingFact)?;
-            let rm_scaled = ctx
-                .scale(&rm_extended, 64, ty.bytes())
-                .ok_or(PccError::Overflow)?;
-            let sum = ctx.add(&rn, &rm_scaled, 64).ok_or(PccError::MissingFact)?;
+            let rn = get_fact_or_fail(vcode, rn)?;
+            let rm = get_fact_or_fail(vcode, rm)?;
+            let rm_extended = fail_if_missing(extend_fact(ctx, rm, extendop))?;
+            let rm_scaled = fail_if_missing(ctx.scale(&rm_extended, 64, ty.bytes()))?;
+            let sum = fail_if_missing(ctx.add(&rn, &rm_scaled, 64))?;
             ctx.check_address(&sum, size as u32)
         }
         &AMode::RegExtended { rn, rm, extendop } => {
-            let rn = vcode.vreg_fact(rn.into()).ok_or(PccError::MissingFact)?;
-            let rm = vcode.vreg_fact(rm.into()).ok_or(PccError::MissingFact)?;
-            let rm_extended = amode_extend(ctx, rm, extendop).ok_or(PccError::MissingFact)?;
-            let sum = ctx
-                .add(&rn, &rm_extended, 64)
-                .ok_or(PccError::MissingFact)?;
+            let rn = get_fact_or_fail(vcode, rn)?;
+            let rm = get_fact_or_fail(vcode, rm)?;
+            let rm_extended = fail_if_missing(extend_fact(ctx, rm, extendop))?;
+            let sum = fail_if_missing(ctx.add(&rn, &rm_extended, 64))?;
             ctx.check_address(&sum, size as u32)
         }
         &AMode::Unscaled { rn, simm9 } => {
-            let rn = vcode.vreg_fact(rn.into()).ok_or(PccError::MissingFact)?;
-            let sum = ctx
-                .offset(&rn, 64, simm9.value as i64)
-                .ok_or(PccError::MissingFact)?;
+            let rn = get_fact_or_fail(vcode, rn)?;
+            let sum = fail_if_missing(ctx.offset(&rn, 64, simm9.value as i64))?;
             ctx.check_address(&sum, size as u32)
         }
         &AMode::UnsignedOffset { rn, uimm12 } => {
-            let rn = vcode.vreg_fact(rn.into()).ok_or(PccError::MissingFact)?;
+            let rn = get_fact_or_fail(vcode, rn)?;
+            // Safety: this will not overflow: `size` should be at
+            // most 32 or 64 for large vector ops, and the `uimm12`'s
+            // value is at most 4095.
             let offset = (uimm12.value as u64) * (size as u64);
```

Comment:
Can we do `a.checked_mul(b).unwrap()` to assert that this will not overflow?

Reply:
Yep, good idea.